mm2src/mm2_metrics/src/mm_metrics.rs

```diff
@@ -0,0 +1,538 @@
+#[macro_export]
+macro_rules! mm_gauge {
+    ($metrics:expr, $name:expr, $value:expr) => {{
+        use $crate::metrics::Recorder;
+        if let Some(recorder) = $crate::recorder::TryRecorder::try_recorder(&$metrics){
+            let key = $crate::metrics::Key::from_static_name($name);
+            let gauge = recorder.register_gauge(&key);
+            gauge.set($value);
+        }
+    }};
+
+    // Register and set gauge with label.
+    ($metrics:expr, $name:expr, $value:expr, $($label_key:expr => $label_val:expr),+) => {{
+        use $crate::metrics::Recorder;
+        if let Some(recorder) = $crate::recorder::TryRecorder::try_recorder(&$metrics){
+            let key = $crate::metrics::Key::from_parts($name, mm_label!($($label_key => $label_val),+).as_slice());
+            let gauge = recorder.register_gauge(&key);
+            gauge.set($value);
+        }
+    }};
+}
+
+/// Update gauge if an MmArc is not dropped yet and metrics system is initialized already.
+#[macro_export]
+macro_rules! mm_timing {
+    ($metrics:expr, $name:expr, $value:expr) => {{
+        use $crate::metrics::Recorder;
+        if let Some(recorder) = $crate::recorder::TryRecorder::try_recorder(&$metrics){
+            let key =$crate::metrics::Key::from_static_name($name);
+            let histo = recorder.register_histogram(&key);
+            histo.record($value);
+        }
+    }};
+
+    // Register and record histogram with label.
+    ($metrics:expr, $name:expr, $value:expr, $($label_key:expr => $label_val:expr),+) => {{
+        use $crate::metrics::Recorder;
+        if let Some(recorder) = $crate::recorder::TryRecorder::try_recorder(&$metrics){
+            let key = $crate::metrics::Key::from_parts($name, mm_label!($($label_key => $label_val),+).as_slice());
+            let histo = recorder.register_histogram(&key);
+            histo.record($value);
+        }
+    }};
+}
+
+/// Market Maker Metrics, used as inner to get metrics data and exporting.
+#[derive(Default, Clone)]
+pub struct Metrics {
+    pub(crate) recorder: Arc<MmRecorder>,
+}
+
+impl Metrics {
+    /// Collect the metrics in Prometheus format.
+    #[cfg(not(target_arch = "wasm32"))]
+    pub fn collect_prometheus_format(&self) -> String { self.recorder.render_prometheus() }
+}
+
+impl MetricsOps for Metrics {
+    fn init(&self) { Metrics::default(); }
+
+    fn init_with_dashboard(&self, log_state: LogWeak, interval: f64) -> MmMetricsResult<()> {
+        let recorder = self.recorder.clone();
+        let runner = TagObserver::log_tag_metrics(log_state, recorder, interval);
+        spawn(runner);
+        Ok(())
+    }
+
+    /// Collect prepared metrics json from the recorder.
+    fn collect_json(&self) -> MmMetricsResult<Value> {
+        serde_json::to_value(self.recorder.prepare_json()).map_to_mm(|err| MmMetricsError::Internal(err.to_string()))
+    }
+}
+
+#[derive(Debug)]
+#[allow(dead_code)]
```

Comment:
Is it required to mark it as `allow(dead_code)`?

Reply:
had it there some reasons seems it's not needed anymore will remove now